### PR TITLE
[Xcode] Support multiple PGO profiles by matching the platform target triple

### PR DESCRIPTION
**Pull request by emw-apple**
#### 37dab5e553f6d157683732b971b5ea7ae830ec66
<pre>
[Xcode] Support multiple PGO profiles by matching the platform target triple
<a href="https://bugs.webkit.org/show_bug.cgi?id=253691">https://bugs.webkit.org/show_bug.cgi?id=253691</a>
rdar://106459115

Reviewed by Alexey Proskuryakov.

Change copy-profiling-data to find a profile to use based on the active
target triple. Using different profiles for different targets is
valuable for making future performance improvements and addressing
backwards compatibility.

Since we may not have a profile checked in for an exact target triple,
at build time, we pick the profile for the closest triple we can find.
For example, for the target triple arm64e-apple-ios16.4:

1. If profiles are available for arm64e-apple-ios16.4, use them.
2. Otherwise, find a profile for a prior version of the same platform
   (e.g. arm64e-apple-ios16.0).
3. If nothing is available for the platform, use the latest macOS
   profile.
4. If WebKitAdditions is unavailable, fall back to the empty profile.

The existing behavior (where all targets using profiles generated for
macOS) falls under case 3. Compatibility issues with downlevels are
handled by cases 1 and 2. And the open-source build (with no profiling
data) is case 4.

* Configurations/WebKitProjectPaths.xcconfig: Fix
  WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH to work correctly in
  engineering builds. Add WK_PROFILING_DATA_FOLDER_PATH to refer to the
  Profiles directory. This needs to be a discrete build setting so that
  it can be unset in open-source builds, to satisfy XCBuild&apos;s dependency
  validation.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Remove
  PROFILE_DATA_PATH variable.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Change
  &quot;Copy Profiling Data&quot; build phase to depend on the entire Profiling/
  directory node. WebKitAdditions is changing to declare a directory
  reference for this path, rather than declaring individual file
  references for the profiles. Any change to the recursive contents of
  the directory will re-run the script.

  Note that Xcode currently has no way of knowing that the (copied)
  profile is an input of all compiler tasks, so a latent incremental
  build bug remains. We should adopt the CLANG_USE_OPTIMIZATION_PROFILE
  setting in a follow-up.
* Source/JavaScriptCore/Scripts/copy-profiling-data.sh: The search logic
  described above.
* Source/WebCore/Configurations/WebCore.xcconfig: Ditto xcconfig
  changes.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Ditto pbxproj
  changes.
* Source/WebKit/Configurations/BaseTarget.xcconfig: Ditto xcconfig
  changes.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Ditto pbxproj changes.
* Tools/Scripts/build-and-collect-pgo-profiles: Approximate the target
  triple we built for by analyzing the build products being tested. This
  is currently very Mac-specific, but this whole script woudl need to be
  changed dramatically to support collection on embedded platforms.

Canonical link: <a href="https://commits.webkit.org/261593@main">https://commits.webkit.org/261593@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/11002b3f9d8a0f94b8b0c55dbbbef26048cb4f11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112202 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21336 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/828 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4003 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120847 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12421 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4764 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117965 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16852 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100035 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105263 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98809 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/576 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45858 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100608 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13744 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/616 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11867 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14422 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101952 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52615 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31872 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8083 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16216 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109995 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27173 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -9821,6 +9821,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				BC18C4140E16F5CD00B34460 /* JavaScriptCore.h in Headers */,
```
Comment (aproskuryakov) [inline Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj]: This seems unrelated?

File: Source/JavaScriptCore/Scripts/copy-profiling-data.sh [hunk 2]
```diff
@@ -4,12 +4,46 @@ if [ -z "${PROFILE_DATA_FLAGS}" ]; then
...
-input_profdata="${SCRIPT_INPUT_FILE_0}"
-input_profdata_type="$(file -b "${input_profdata}")"
-
+profiles="${SCRIPT_INPUT_FILE_0}"
 fallback_decompressed_profdata="${SCRIPT_INPUT_FILE_1}"
 derived_decompressed_profdata="${SCRIPT_OUTPUT_FILE_0}"
 
+# Search for a PGO profile corresponding to the current target triple or a
+# previous OS version triple. In lieu of an exact match, fall back to the most
+# recent macOS profile.
```
Comment (aproskuryakov) [inline Source/JavaScriptCore/Scripts/copy-profiling-data.sh:16]: I'm not quite sure how this works, given that we need to support very many old versions, which cannot fall back to latest. I guess I'll know once I see the content in WebKitAdditions.

**Review by aproskuryakov — APPROVED** [review]

File: Source/JavaScriptCore/Scripts/copy-profiling-data.sh (hunk 2 above)
Comment (emw-apple) [reply Source/JavaScriptCore/Scripts/copy-profiling-data.sh:16]: Old macOS versions will be handled by the second clause, "a previous OS version triple". Embedded platforms will use the latest Mac profile, like they do now.

File: Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (emw-apple) [reply Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj]: Fallout from running sort-Xcode-project-file, which I don't need to do since I'm not adding/removing files in the project. Fixed.

Comment (emw-apple) [reply Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj]: Ah, actually, it's getting run automatically (via the `pre-commit` hook), because I'm committing changes to the pbxproj. Probably best to just accept the change it's making.

The reason it's doing the reordering in the first place is because of 8cd33528ad547c1bec218e3b05d01e1d354c1b2e, specifically:
> Work around a header-ordering bug in tapi (rdar://104432605) by sorting all umbrella headers to the top of their targets' Headers phase.

**Comment by justinmichaud** [comment]
🎉

**Review by aproskuryakov — APPROVED** [review]
